Pull request test_esr

Pake-TU opened the pull request:
**Summary of improvements**
- Changed the scheduler
- Tests for most of thing.go, esr.go, scheduler.go

Pake-TU committed:
Fixed typos

Pake-TU committed:
Added test for initTemplate()

Pake-TU committed:
Removed unnecessary logic

Pake-TU committed:
Added test for getUniqueSystems()

Pake-TU committed:
Refactored and simplified FilterByServiceDefinitionAndDetails()

Pake-TU committed:
Added help functions etc for testing FilterByServiceDefinitionAndDetails()

Pake-TU committed:
Finished test for FilterByServiceDefinitionAndDetails()

Pake-TU committed:
Added test for newResource

Pake-TU committed:
Added CoreS to createTestSystem()

Pake-TU committed:
Added tests for roleStatus() and peersList()

Pake-TU committed:
Added tests for systemList()

Pake-TU committed:
Changed helpfunctions returned unitasset to pointer

Pake-TU committed:
Fixed linter errors: added error handlers and var declaration

Pake-TU committed:
added missing mutex locks & unlocks

Pake-TU committed:
added first test for serviceRegistryHandler()

Pake-TU committed:
Added tests for serviceRegistryHandler()

lmas committed:
Adds go.mod files as they are required for Go projects

Pake-TU committed:
Added tests for updateDB()

Pake-TU committed:
Added tests for queryDB and cleanDB()

Pake-TU committed:
Fixed linter errors

Pake-TU committed:
Fixed spellchecker errors

Pake-TU committed:
Simplified scheduler

gabaxh commented on esr/esr.go:
```diff
@@ -120,7 +120,9 @@ func (ua *UnitAsset) Serving(w http.ResponseWriter, r *http.Request, servicePath
 		w.WriteHeader(http.StatusServiceUnavailable)
-		w.Write([]byte("Service Unavailable"))
+		if _, err := w.Write([]byte("Service Unavailable")); err != nil {
+			log.Printf("error occurreded while writing to responsewriter: %v", err)
```
Spelling mistake, pretty sure at least

Pake-TU replied:
You're right, thanks for the heads up 😆

gabaxh commented on esr/esr.go:
```diff
@@ -203,22 +205,30 @@ func (ua *UnitAsset) queryDB(w http.ResponseWriter, r *http.Request) {
 		case servvicesList := <-recordsRequest.Result:
```
Should it be "servvicesList" and not "servicesList"?

gabaxh commented on esr/esr.go:
```diff
@@ -203,22 +205,30 @@ func (ua *UnitAsset) queryDB(w http.ResponseWriter, r *http.Request) {
 			text := "<!DOCTYPE html><html><body>"
-			w.Write([]byte(text))
+			if _, err := w.Write([]byte(text)); err != nil {
+				log.Printf("error occurreded while writing to responsewriter: %v", err)
```
Spelling again

gabaxh commented on esr/esr.go:
```diff
@@ -203,22 +205,30 @@ func (ua *UnitAsset) queryDB(w http.ResponseWriter, r *http.Request) {
-			w.Write([]byte(text))
-			for _, serRec := range servvicesList {
+			if _, err := w.Write([]byte(text)); err != nil {
+				log.Printf("error occurreded while writing to responsewriter: %v", err)
```
And again, seems like spellcheck does not pick up on these unfortunately.

gabaxh commented on esr/esr.go:
```diff
@@ -203,22 +205,30 @@ func (ua *UnitAsset) queryDB(w http.ResponseWriter, r *http.Request) {
-				w.Write([]byte(fmt.Sprintf("<li>%s</li>", sLine)))
+				sLine := "<p>Service ID: " + strconv.Itoa(int(servRec.Id)) + " with definition <b><a href=\"" + hyperlink + "\">" + servRec.ServiceDefinition + "</b></a> from the <b>" + servRec.SystemName + "/" + uaName + "</b> with details " + metaservice + " will expire at: " + servRec.EndOfValidity + "</p>"
+				if _, err := w.Write([]byte(fmt.Sprintf("<li>%s</li>", sLine))); err != nil {
+					log.Printf("error occurreded while writing to responsewriter: %v", err)
```
Here too

gabaxh commented on esr/esr.go:
```diff
@@ -203,22 +205,30 @@ func (ua *UnitAsset) queryDB(w http.ResponseWriter, r *http.Request) {
 			text = "</ul></body></html>"
-			w.Write([]byte(text))
+			if _, err := w.Write([]byte(text)); err != nil {
+				log.Printf("error occurreded while writing to responsewriter: %v", err)
```
Here

gabaxh commented on esr/esr.go:
```diff
@@ -203,22 +205,30 @@ func (ua *UnitAsset) queryDB(w http.ResponseWriter, r *http.Request) {
+			}
 		case <-time.After(5 * time.Second): // Optional timeout
 			http.Error(w, "Request timed out", http.StatusGatewayTimeout)
 			log.Println("Failure to process service listing request")
```
On line 237 orchestrator in the comments is wrong,

gabaxh commented on esr/esr.go:
```diff
@@ -339,7 +349,9 @@ func (ua *UnitAsset) roleStatus(w http.ResponseWriter, r *http.Request) {
 		w.WriteHeader(http.StatusServiceUnavailable)
-		w.Write([]byte("Service Unavailable"))
+		if _, err := w.Write([]byte("Service Unavailable")); err != nil {
+			log.Printf("error occurreded while writing to responsewriter: %v", err)
```
Here too

gabaxh commented on esr/scheduler.go:
```diff
@@ -16,132 +16,43 @@
+	taskMap map[int]*time.Timer // list elements has id, timer
 }
 
 // NewScheduler creates a new scheduler
```
Kinda useless comment, the comments says exactly what the function name says.

gabaxh commented on esr/scheduler.go:
```diff
@@ -16,132 +16,43 @@
-
-// Scheduler struct type with the list and two channels
+// Scheduler struct type with the list and three channels
 type Scheduler struct {
```
Amazing how much simpler the scheduler now is, great work!

Pake-TU committed:
Added tests for scheduler

Pake-TU committed:
Fixed some errorhandlers, spellings and locks in newResource

Pake-TU committed:
fixed test after changes in esr.go

Pake-TU committed:
Fixed comments and removed unused struct

Pake-TU committed:
Added delete test for serviceRegistryHandler()

Pake-TU committed:
Added missing http.Error() in error handlers for updateDB()

Pake-TU committed:
fixed errorhandlers in esr.go, and added testcase in esr_test.go

lmas commented on esr/esr.go:
Skipping review on this file.

lmas commented on esr/thing.go:
Skipping review on this file.

lmas commented on esr/esr_test.go:
```diff
@@ -0,0 +1,614 @@
+	var serviceAmount int
+	for x := range 5 {
+		serviceAmount++
+		ua.serviceRegistry[x] = forms.ServiceRecord_v1{Id: x, SystemName: fmt.Sprintf("testSys%d", x),
```
This is a case where I think it's better to split up each field on a separate line, to improve readability and seeing quickly how each field is set.

lmas commented on esr/esr_test.go:
```diff
@@ -0,0 +1,614 @@
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+				sendAddRequest(0, "test", "testPath", "", ua.requests)
+
+				return w, r
+			},
+			"Bad case, request returns error",
+		},
+		{
+			200,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+
+				rec := &forms.ServiceRecord_v1{
+					Id:      0,
+					Version: "ServiceRecord_v1",
+				}
+
+				data, _ := json.Marshal(rec)
+				w = httptest.NewRecorder()
+				body := io.NopCloser(bytes.NewReader(data))
+				r = httptest.NewRequest(http.MethodPost, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Good case, everything passes",
+		},
+		{
+			200,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+				w = httptest.NewRecorder()
+				body := io.NopCloser(strings.NewReader(""))
+				r = httptest.NewRequest(http.MethodGet, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Good case, default case",
+		},
```
These setup functions all looks very similar. Would be better to break out `ua.leading` and the body and use them in the table instead of the setup func. The two cases where you're doing something different in the setup could probably be refactored or handled as separate special cases.

lmas commented on esr/esr_test.go:
```diff
@@ -0,0 +1,614 @@
+				ua.leading = true
+
+				w = httptest.NewRecorder()
+				body := io.NopCloser(strings.NewReader(`{"id": 0, "version":"SignalA_v1.0"}`))
+				r = httptest.NewRequest(http.MethodPost, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Bad case POST, request returns error",
+		},
+		{
+			http.StatusOK,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+
+				w = httptest.NewRecorder()
+				body := io.NopCloser(strings.NewReader(`{"id": 0, "version":"ServiceQuest_v1"}`))
+				r = httptest.NewRequest(http.MethodPost, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Good case POST, request returns a result",
+		},
+		{
+			http.StatusMethodNotAllowed,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+
+				w = httptest.NewRecorder()
+				body := io.NopCloser(strings.NewReader(`{"id": 0, "version":"ServiceQuest_v1"}`))
+				r = httptest.NewRequest(http.MethodDelete, "http://localhost/reg", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Bad case default, unsupported http method",
+		},
+	}
```
Same here, don't abuse the setup functions. If they are doing the same thing anyway, might as well put it in the test function directly.

lmas commented on esr/esr_test.go:
```diff
@@ -0,0 +1,614 @@
+
+func TestCleanDB(t *testing.T) {
+	params := []cleanDBParams{
+		{
+			http.StatusBadRequest,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+
+				w = httptest.NewRecorder()
+				body := io.NopCloser(strings.NewReader(`{"id": 0, "version":"ServiceQuest_v1"}`))
+				r = httptest.NewRequest(http.MethodDelete, "http://localhost/reg/a", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Bad case DELETE, couldn't convert id to int",
+		},
+		{
+			200,
+			func(ua *UnitAsset) (w *httptest.ResponseRecorder, r *http.Request) {
+				ua.leading = true
+
+				w = httptest.NewRecorder()
+				body := io.NopCloser(strings.NewReader(`{"id": 0, "version":"ServiceQuest_v1"}`))
+				r = httptest.NewRequest(http.MethodGet, "http://localhost/reg/a", body)
+				r.Header = map[string][]string{"Content-Type": {"application/json"}}
+
+				return w, r
+			},
+			"Bad case default, unsupported http method",
+		},
+	}
```
Same here.

lmas commented on esr/scheduler_test.go:
```diff
@@ -0,0 +1,61 @@
+// Tests for scheduler
+// -------------------- //
+
+func TestAddTask(t *testing.T) {
+	sched := NewScheduler()
+	now := time.Now()
+
+	// Add the task
+	sched.AddTask(now.Add(25*time.Second), func() {}, 0)
+
+	if _, exists := sched.taskMap[0]; !exists {
+		t.Errorf("Task was not present")
+	}
+}
```
This could be extended with more cases, for example:
- Will new tasks be inserted before/after already existing tasks in the correct order?
- What happens with tasks with bad times (negatives?) ?
- What happens with tasks with bad IDs (negatives?) ?
- What happens with duplicate tasks IDs?

Pake-TU replied:
- Added test to ensure chronological order of tests
- Bad times unhandled, we assume it's been handled by the caller
- Bad IDs (negatives) unhandled, we assume it's been handled by caller, since id is linked to a service in the serviceMap
- Added test for when same ID is repeated before timer is up (updated timer and/or job)

lmas replied:
Yap, agreed after our long discussions.

lmas commented on esr/scheduler_test.go:
```diff
@@ -0,0 +1,61 @@
+	}
+}
+
+func TestRemoveTask(t *testing.T) {
+	// Case: task exists
+	sched := NewScheduler()
+	now := time.Now()
+
+	// Add the task and then remove it
+	sched.AddTask(now.Add(25*time.Second), func() {}, 0)
+	sched.RemoveTask(0)
+
+	if _, exists := sched.taskMap[0]; exists {
+		t.Errorf("Unexpected task exists in taskMap[0]")
+	}
+
+	// Case: task doesn't exist
+	sched = NewScheduler()
+	// Add the task and then remove it
+	sched.RemoveTask(0)
+
+	if _, exists := sched.taskMap[0]; exists {
+		t.Errorf("Unexpected task exists in taskMap[0]")
+	}
+}
```
Same here, add more test cases. What happens with multiple tasks, will the order remain sorted?

Pake-TU replied:
This is handeled by time.AfterFunc()

lmas replied:
As has been discussed yap. I'm liking the extra return check

lmas commented on esr/scheduler.go:
```diff
@@ -16,132 +16,44 @@
+// Stop() loops through the task map and turns off the timer for each tasks job
 func (s *Scheduler) Stop() {
-	s.stopChan <- struct{}{}
+	for _, value := range s.taskMap {
```
"value" isn't very descriptive. In the other funcs you've called it "timer" instead, which I think fits better and remains conventional.

lmas commented on esr/scheduler_test.go:
```diff
@@ -0,0 +1,61 @@
+	}
+}
+
+func TestStop(t *testing.T) {
+	sched := NewScheduler()
+	now := time.Now()
+
+	// Add some tasks and make sure Stop() works as intended
+	sched.AddTask(now.Add(25*time.Second), func() {}, 0)
+	sched.AddTask(now.Add(25*time.Second), func() {}, 1)
+	sched.AddTask(now.Add(25*time.Second), func() {}, 2)
+	sched.AddTask(now.Add(25*time.Second), func() {}, 3)
+	sched.Stop()
+
+	if len(sched.taskMap) > 0 {
+		t.Errorf("Expected taskMap to be empty, has %d keys", len(sched.taskMap))
+	}
+}
```
I think it would be good to try to test that each task timer was stopped properly too, if it's possible.

Pake-TU replied:
I don't know how that would work, since the `Stop()` function for Schedulers loops through its taskMap and stops the timer of every task, and then resets the map.

~I'd have to rethink the current approach, maybe save all the timers in a list and return that list to check if the timer's been stopped?~
EDIT: After some discussion, we've come up with a good solution

lmas replied:
Simple and effective!

lmas commented on esr/thing_test.go:
```diff
@@ -0,0 +1,729 @@
+// Help functions and structs to tests initTemplate()
+// ----------------------------------------------------- //
+
+func TestInitTemplate(t *testing.T) {
+	expectedServices := []string{"register", "query", "unregister", "status"}
+
+	ua := initTemplate()
+	services := ua.GetServices()
+
+	// Check if expected name and services are present
+	if ua.GetName() != "registry" {
+		t.Errorf("Name mismatch expected 'registry', got: %s", ua.GetName())
+	}
+
+	for _, s := range expectedServices {
+		if _, ok := services[s]; !ok {
+			t.Errorf("Expected service '%s' to be present", s)
+		}
+	}
+}
```
As I said to Gabriel, skip testing `initTemplate` and `newResource` unless they're trying to do anything fancy. Testing if a function sets a couple of fields in a struct (that has a high probability of changing in the near future) doesn't provide any real value and has an increased risk of adding problems.

lmas commented on esr/thing_test.go:
```diff
@@ -0,0 +1,729 @@
+	testCase  string
+}
+
+func TestNewResource(t *testing.T) {
+	params := []newResourceParams{
+		{
+			func() (sys components.System) { return createTestSystem() },
+			func() (confAsset usecases.ConfigurableAsset) { return createConfAssetBrokenTraits() },
+			"Case: unmarshal traits fails",
+		},
+		{
+			func() (sys components.System) { return createTestSystem() },
+			func() (confAsset usecases.ConfigurableAsset) { return createConfAssetMultipleTraits() },
+			"Case: confAsset has multiple traits",
+		},
+	}
+
+	for _, c := range params {
+		sys := c.setup()
+		uac := c.confAsset()
+
+		ua, shutdown := newResource(uac, &sys)
+		shutdown()
+		if ua.GetName() != "testRegistrar" {
+			t.Errorf("Name mismatch, expected '%s' got '%s'", uac.Name, ua.GetName())
+		}
+	}
+}
```
The `newResource` is just wonky atm, skip testing it even if it's doing more than just setting fields. Not worth chasing coverage percent for a function that risk breaking in a month.

lmas commented on esr/thing_test.go:
```diff
@@ -0,0 +1,729 @@
+		{
+			true,
+			func(ua *UnitAsset) error {
+				sendAddRequest(0, "testDef", "subP", time.Now().Format(time.RFC3339), ua.requests)
```
Any good reason for not testing the error from this call?

Pake-TU replied:
I figured since it passed in the first test case it would be fine not to check it for errors in the later test cases, since it's always the first request sent.

I've added error handlers for it anyway, just to be on the safe side

lmas replied:
Thinks could (theoretically) change between cases.

lmas commented on esr/thing_test.go:
```diff
@@ -0,0 +1,729 @@
+		{
+			true,
+			func(ua *UnitAsset) error {
+				sendAddRequest(0, "testDef", "subP", time.Now().Format(time.RFC3339), ua.requests)
```
Any good reason for not testing the error from this call?

Pake-TU replied:
see https://github.com/lmas/mbaigo_systems/pull/3#discussion_r2225117820

lmas commented on esr/thing_test.go:
```diff
@@ -0,0 +1,729 @@
+		{
+			true,
+			func(ua *UnitAsset) error {
+				sendAddRequest(0, "testDef", "subP", time.Now().Format(time.RFC3339), ua.requests)
```
Any good reason for not testing the error from this call?

Pake-TU replied:
see https://github.com/lmas/mbaigo_systems/pull/3#discussion_r2225117820

lmas commented on esr/thing_test.go:
```diff
@@ -0,0 +1,729 @@
+			true,
+			func(ua *UnitAsset) error {
+				ch := ua.requests
+				sendAddRequest(0, "testDef", "subP", time.Now().Format(time.RFC3339), ch)
```
Any good reason for not testing the error from this call?

Pake-TU replied:
see https://github.com/lmas/mbaigo_systems/pull/3#discussion_r2225117820

lmas commented on esr/thing_test.go:
```diff
@@ -0,0 +1,729 @@
+			false,
+			func(ua *UnitAsset) error {
+				ch := ua.requests
+				sendAddRequest(0, "testDef", "subP", time.Now().Format(time.RFC3339), ch)
```
Any good reason for not testing the error from this call?

Pake-TU replied:
see https://github.com/lmas/mbaigo_systems/pull/3#discussion_r2225117820

lmas commented on esr/thing_test.go:
```diff
@@ -0,0 +1,729 @@
+			false,
+			func(ua *UnitAsset) error {
+				ch := ua.requests
+				sendAddRequest(0, "testDef", "subP", time.Now().Format(time.RFC3339), ch)
```
Any good reason for not testing the error from this call?

Pake-TU replied:
see https://github.com/lmas/mbaigo_systems/pull/3#discussion_r2225117820

Pake-TU committed:
Added functioncall in checkExpiration() to remove task from scheduler

Pake-TU committed:
Fixed PR comments

Pake-TU committed:
changed errorhandlers in testcases

Pake-TU committed:
Fixed PR comments

Pake-TU committed:
Added returns for RemoveTask() and Stop()

Pake-TU committed:
Finished tests, and comments from PR

lmas committed:
Small fixes from the review

lmas reviewed (approved):
LGTM, waiting for a :+1: from @gabaxh

Pake-TU committed:
Moved and slightly changed the ID check to ensure an old service will re-register with a new ID if registrar restarts

Pake-TU committed:
Merge branch 'test_esr' of github.com:lmas/mbaigo_systems into test_esr

gabaxh reviewed (approved):
Yeah, looks good to me!

lmas commented:
@Pake-TU last merge broke a test:
```
--- FAIL: TestServiceRegistryHandlerAdd (0.02s)
    thing_test.go:273: Expected errors in 'Bad case, service doesn't exist in registry'
```

Linter warning too:
```
> staticcheck . 
thing.go:200:33: struct type 'github.com/sdoque/systems/esr.Traits' doesn't have any exported fields, nor custom marshaling (SA9005)
```

Pake-TU committed:
Removed unnecessary testcase

Pake-TU committed:
Fixed linter error

Pake-TU commented:
> @Pake-TU last merge broke a test:
> 
> ```
> --- FAIL: TestServiceRegistryHandlerAdd (0.02s)
>     thing_test.go:273: Expected errors in 'Bad case, service doesn't exist in registry'
> ```
> 
> Linter warning too:
> 
> ```
> > staticcheck . 
> thing.go:200:33: struct type 'github.com/sdoque/systems/esr.Traits' doesn't have any exported fields, nor custom marshaling (SA9005)
> ```

- Fixed test error by removing the test case, since fixing an issue it's no longer needed.
- Fixed linter error by moving traits (and deleting the unmarshaltraits func) into the struct it's supposed to be embedded into, since they're private (wont unmarshal) and isn't supposed to be changed by hand.
